streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -251,26 +252,69 @@ public StreamTask(final TaskId id,
         if (eosEnabled) {
             initializeTransactions();
         }
-    }
 
-    @Override
-    public boolean initializeStateStores() {
-        log.debug("Initializing state stores");
+        try {
+            final Map<TopicPartition, OffsetAndMetadata> offsetsAndMetadata = consumer.committed(partitions);
+            initializeCommittedOffsets(offsetsAndMetadata);
+            initializeTaskTime(offsetsAndMetadata);
+        } catch (final AuthorizationException e) {
+            throw new ProcessorStateException(String.format("task [%s] AuthorizationException when initializing offsets for %s", id, partitions), e);
+        } catch (final WakeupException e) {
+            throw e;
+        } catch (final KafkaException e) {
+            throw new ProcessorStateException(String.format("task [%s] Failed to initialize offsets for %s", id, partitions), e);
+        }
+    }
 
+    private void initializeCommittedOffsets(final Map<TopicPartition, OffsetAndMetadata> offsetsAndMetadata) {
         // Currently there is no easy way to tell the ProcessorStateManager to only restore up to
         // a specific offset. In most cases this is fine. However, in optimized topologies we can
         // have a source topic that also serves as a changelog, and in this case we want our active
         // stream task to only play records up to the last consumer committed offset. Here we find
         // partitions of topics that are both sources and changelogs and set the consumer committed
         // offset via stateMgr as there is not a more direct route.
         final Set<String> changelogTopicNames = new HashSet<>(topology.storeToChangelogTopic().values());
-        final Set<TopicPartition> sourcePartitionsAsChangelog = new HashSet<>(partitions)
-            .stream().filter(tp -> changelogTopicNames.contains(tp.topic())).collect(Collectors.toSet());
-        final Map<TopicPartition, Long> committedOffsets = committedOffsetForPartitions(sourcePartitionsAsChangelog);
-        stateMgr.putOffsetLimits(committedOffsets);
+        final Map<TopicPartition, Long> committedOffsetsForChangelogs = offsetsAndMetadata
+            .entrySet()
+            .stream()
+            .filter(e -> changelogTopicNames.contains(e.getKey().topic()))
+            .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().offset()));
 
-        registerStateStores();
+        // those do not have a committed offset would default to 0
+        for (final TopicPartition tp : partitions) {
+            committedOffsetsForChangelogs.putIfAbsent(tp, 0L);
+        }
+
+        stateMgr.putOffsetLimits(committedOffsetsForChangelogs);
+    }
+
+    private void initializeTaskTime(final Map<TopicPartition, OffsetAndMetadata> offsetsAndMetadata) {
```

Comment:
This is not private. Same implementation as before, however, we get `offsetsAndMetadata` passed in, and don't use the consumer to get them from the brokers.